src/goto-harness/memory_snapshot_harness_generator.h

```diff
@@ -275,11 +275,68 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
   /// \return pointer depth of type \p t
   size_t pointer_depth(const typet &t) const;
 
-  /// Recursively test pointer reference
-  /// \param expr: expression to be tested
-  /// \param name: name to be located
-  /// \return true if \p expr refers to an object named \p name
-  bool refers_to(const exprt &expr, const irep_idt &name) const;
+  void collect_references(const exprt &expr, std::set<irep_idt> &reference_set)
+    const;
+
+  /// Simple structure for linearising posets. Should be constructed with a map
+  /// assigning a key to a set of keys that precede it.
+  template <typename Key, typename Value>
+  struct partial_ordert
+  {
+  public:
+    explicit partial_ordert(const std::map<Key, std::set<Key>> &por) : por(por)
+    {
+    }
+
+    void sort(
+      const std::vector<std::pair<Key, Value>> &input,
+      std::vector<std::pair<Key, Value>> &output)
+    {
+      std::map<Key, Value> searchable_input;
+      for(const auto &item : input)
+        searchable_input[item.first] = item.second;
+
+      for(const auto &item : input)
+        recurse_then_insert(item.first, searchable_input, output);
+    }
+
+  private:
+    const std::map<Key, std::set<Key>> &por;
+
+    std::set<Key> inserted;
+    std::set<Key> seen;
+
+    void recurse_then_insert(
+      const Key &key,
+      const std::map<Key, Value> &input,
+      std::vector<std::pair<Key, Value>> &output)
+    {
```

Comment:
This seems to be basically doing a DFS. Might be worth just renaming the function to reflect that.

It could just have an API like

```
template<typename Value, typename GetNodeFromValue, typename GetNeighbors, typename OnFirstVisit>
void depth_first_search(Value &&entry_node, GetKeyFromValue &&get_key_from_value, GetNeighbors &&getNeighbors, OnFirstVisit &&onFirstVisit);
```

and then you could use it above like
```
auto get_key_from_value = [](const std::pair<Key, Value>& kv_pair) { return kv_pair.first; };
auto get_neighbors = [&](const Key& key) { return por.at(key); }
auto on_first_visit = [&](const Value& value) { output.push_back(value); }
for(const auto &item : input) {
  dfs(item, get_key_from_value, get_neighbors, on_first_visit);
}
```

(Note that I've changed `Value` to what was previously `std::pair<Key,Value>`, and instead there's a separate function for getting the key from the value).

:exclamation: This is just a suggestion, I'm not saying you have to do it exactly this way obviously, but in any case I think it's worth emphasising that "recurse then insert" is basically doing DFS in the code, either by renaming the function or mentioning it in a comment

Reply:
I went half-way. Given that the ordering is on `keys` only, the structure is only templated by `Key` and the `sort`, `dfs` functions and templated by some `T`.